sdt_probe_offset.c

```diff
@@ -26,346 +26,389 @@
 
 /*
  * Default value of MAXSTRINGLEN in SystemTap is 128 but this value can be
  * changed by the user.
  */
-#define MAX_STR_LEN 128
-#define ARRAY_LEN(a) sizeof(a)/sizeof(a[0])
+#define MAX_STR_LEN	128
+#define ARRAY_LEN(a)	(sizeof(a) / sizeof(a[0]))
 
 /*
- * Convert the virtual address in binary to the offset of the instruction in the
- * binary file.
- * Returns the offset on success,
- * Returns -1 in case of failure
+ * Convert the virtual address in a binary's mapping to the offset of
+ * the corresponding instruction in the binary file.
+ *
+ * Returns the offset on success or -1 in case of failure.
  */
 static long convert_addr_to_offset(Elf *elf_handle, size_t addr)
 {
 	long ret;
-	int text_section_found;
+	bool text_section_found = false;
 	size_t text_section_offset, text_section_addr, offset_in_section;
 	char *section_name;
 	size_t section_idx;
-	Elf_Scn *elf_section;
+	Elf_Scn *elf_section = NULL;
 	GElf_Shdr elf_section_hdr;
 	
 	if (!elf_handle) {
 		fprintf (stderr, "Invalid ELF handle.\n");
 		ret = -1; 
-		goto err;
+		goto error;
 	}	
 
 	ret = elf_getshdrstrndx(elf_handle, &section_idx);
 	if (ret) {
-		fprintf(stderr, "ELF get header index failed: %s.\n", elf_errmsg(-1));
+		fprintf(stderr, "ELF get header index failed: %s.\n",
+				elf_errmsg(-1));
 		ret = -1; 
-		goto err;
+		goto error;
 	}
 
-	elf_section = NULL;
-	text_section_found = 0;
-
 	while((elf_section = elf_nextscn(elf_handle, elf_section)) != NULL) {
 		if (gelf_getshdr(elf_section, &elf_section_hdr) == NULL) {
 			fprintf(stderr, "GELF get section header failed: %s.\n",
 					elf_errmsg(-1));
 			ret = -1;
-			goto err;
+			goto error;
 		}
 
-		section_name = elf_strptr(elf_handle, section_idx, elf_section_hdr.sh_name);
+		section_name = elf_strptr(elf_handle, section_idx,
+				elf_section_hdr.sh_name);
 		if (section_name == NULL) {
 			fprintf(stderr, "ELF retrieve string pointer failed: %s.\n",
 					elf_errmsg(-1));
 			ret = -1;
-			goto err;
+			goto error;
 		}
 
-		if (strncmp(section_name, TEXT_SECTION_NAME, sizeof(TEXT_SECTION_NAME)) == 0) {
+		if (strncmp(section_name, TEXT_SECTION_NAME,
+				sizeof(TEXT_SECTION_NAME)) == 0) {
 			text_section_offset = elf_section_hdr.sh_offset;
 			text_section_addr = elf_section_hdr.sh_addr;
-			text_section_found = 1;
+			text_section_found = true;
 			break;
 		}
 	}
 
 	if (!text_section_found) {
 		fprintf(stderr, "Text section not found in binary.\n");
 		ret = -1;
-		goto err;
+		goto error;
 	}
 
 	/*
 	 * To find the offset of the addr from the beginning of the .text
 	 * section.
 	 */
+	if (text_section_addr > addr) {
+		fprintf(stderr, "Invalid section address found: addr=%zux, text_section_addr=%zux.\n",
+				addr, text_section_addr);
+		ret = -1;
+		goto error;
+	}
 	offset_in_section = addr - text_section_addr;
 			      
 	/*
 	 * Add the offset in the section to the offset of the section from the
-	 * beginning of the binary.
+	 * beginning of the binary's file.
```

Comment:
I used "binary's file" meaning the file associated with the binary as opposed to the "mapped" binary... But realistically, "of the binary file" is not ambiguous.